### PR TITLE
Add MNIST dataset & drop torchvision dep. from tests

### LINKED ISSUE #942: create own MNIST dataset
## 🚀 Feature

We shall get free of `torchvision` even in tests we wound be impacted by they optional crashes

### Motivation

for example `torchvision` is not on pip for python 3.8
also particular `torchvision` enforce `torch` versions

### Pitch

we reimplement MNIST dataset and class downloading data from own web
Extensions:
- define to use just a subset
- define used digits
- etc.

### Additional context

discussed in #859 and #917 and https://github.com/PyTorchLightning/pytorch-lightning/pull/917#issuecomment-590087298

Resources:
- https://pl-public-data.s3.amazonaws.com/MNIST/processed/test.pt
- https://pl-public-data.s3.amazonaws.com/MNIST/processed/training.pt
- https://pl-public-data.s3.amazonaws.com/MNIST/raw/t10k-images-idx3-ubyte
- https://pl-public-data.s3.amazonaws.com/MNIST/raw/t10k-images-idx3-ubyte.gz
- https://pl-public-data.s3.amazonaws.com/MNIST/raw/t10k-labels-idx1-ubyte
- https://pl-public-data.s3.amazonaws.com/MNIST/raw/t10k-labels-idx1-ubyte.gz
- https://pl-public-data.s3.amazonaws.com/MNIST/raw/train-images-idx3-ubyte
- https://pl-public-data.s3.amazonaws.com/MNIST/raw/train-images-idx3-ubyte.gz
- https://pl-public-data.s3.amazonaws.com/MNIST/raw/train-labels-idx1-ubyte
- https://pl-public-data.s3.amazonaws.com/MNIST/raw/train-labels-idx1-ubyte.gz

### PR DESCRIPTION
**Pull request by awaelchli**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #942 and follow up to #990.

- Added a custom MNIST class that does not rely on torchvision. 
- Added download links to MNIST on S3 
- Removed torchvision dependency from tests
- PL examples still use the real torchvision

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
why is it trying to run the imagenet example? It's not part of the tests, or is it?

**Comment by awaelchli** [comment]
> just curious, do we need all these functions/methods, I would rather use the minimal version so we do not maintain it in future...

ok, I will try to compress everything. Most of the code is for downloading and extracting the dataset files. 

I didn't want to remove any tests, that's why I tried to override the imports. Maybe you go ahead and do the splitting of examples and tests first, then I will adjust my changes to that.

**Comment by Borda** [comment]
> ok, I will try to compress everything. Most of the code is for downloading and extracting the dataset files.

Well I didn't check it in detail... 

> I didn't want to remove any tests, that's why I tried to override the imports. Maybe you go ahead and do the splitting of examples and tests first, then I will adjust my changes to that.

You are right, we shall not change any test, just swap dataset source lok

**Comment by awaelchli** [comment]
@Borda I made it a lot simpler. It is not necessary to do all the data extraction that torchvision does. The MNIST class is now very simple as you wished. And I got rid of the transforms, so we can even drop the PIL requirement now. No PIL to Tensor conversions needed. 

Now we just need to decouple the tests from examples :)

**Comment by Borda** [comment]
@awaelchli I have prepared the split, check #990

**Comment by awaelchli** [comment]
@Borda Great! I think yours should be merged first, because I can't make the tests pass otherwise.

File: .gitignore [hunk 1]
```diff
@@ -130,5 +129,4 @@ ENV/
 
 # data
 .data/
-datasets/
```
Comment (Borda) [inline .gitignore]: add this back

File: tests/datasets/mnist.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            downloaded again.
+    """
+
+    resources = [
+        "https://pl-public-data.s3.amazonaws.com/MNIST/processed/training.pt",
+        "https://pl-public-data.s3.amazonaws.com/MNIST/processed/test.pt",
+    ]
+
+    training_file = 'training.pt'
+    test_file = 'test.pt'
```
Comment (Borda) [inline tests/datasets/mnist.py]: rather CAPITAL

File: tests/datasets/mnist.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        if download:
+            self.download()
+
+        if not self._check_exists():
+            raise RuntimeError('Dataset not found.')
+
+        if self.train:
+            data_file = self.training_file
+        else:
+            data_file = self.test_file
```
Comment (Borda) [inline tests/datasets/mnist.py]:
```suggestion
        data_file = self.training_file if self.train else self.test_file
```

File: .gitignore (hunk 1 above)
Comment (awaelchli) [reply .gitignore]: mnist.py is in there, it will get ignored. 
the gitignore already contains "mnist/", is that not enought? I think we don't use any other datasets in test. where should I move mnist.py then?

Comment (Borda) [reply .gitignore]: so what about moving the mnist.py to model directly... I have usually associated `datasets/` with a folder with stored datasets

**Comment by Borda** [comment]
FYI, https://github.com/pytorch/vision/issues/1938#issuecomment-594497401

File: tests/datasets/mnist.py (hunk 2 above)
Comment (awaelchli) [reply tests/datasets/mnist.py]: also underscore?

Comment (Borda) [reply tests/datasets/mnist.py]: I would make it as `TRAIN_FILE_NAME` and `TEST_FILE_NAME`

**Comment by Borda** [comment]
it seems that `torchvision` won't recover any time soon, so the only way is our simple dataset...
@williamFalcon any other idea?

File: .gitignore (hunk 1 above)
Comment (awaelchli) [reply .gitignore]: ok done

File: tests/models/mnist.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
+import os
```
Comment (Borda) [inline tests/models/mnist.py]: what about name it `datasets.py` and move also `TestingMNIST` here so all datasets are in one place

Comment (awaelchli) [reply tests/models/mnist.py]: good idea

**Comment by Borda** [comment]
Thx for your patience, it was quite hectic with the last release, but this shall be solved also soon

**Comment by awaelchli** [comment]
@Borda unblocked and ready for review :)

File: tests/base/datasets.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    https://github.com/pytorch/vision/blob/build/v0.5.0/torchvision/datasets/mnist.py
+
+    Args:
+        root (string): Root directory of dataset where ``MNIST/processed/training.pt``
+            and  ``MNIST/processed/test.pt`` exist.
+        train (bool, optional): If True, creates dataset from ``training.pt``,
+            otherwise from ``test.pt``.
+        download (bool, optional): If true, downloads the dataset from the internet and
+            puts it in root directory. If dataset is already downloaded, it is not
+            downloaded again.
```
Comment (Borda) [inline tests/base/datasets.py]:
```suggestion
        root: Root directory of dataset where ``MNIST/processed/training.pt``
            and  ``MNIST/processed/test.pt`` exist.
        train: If ``True``, creates dataset from ``training.pt``,
            otherwise from ``test.pt``.
        normalize: mean and std deviation of the MNIST dataset.
        download: If true, downloads the dataset from the internet and
            puts it in root directory. If dataset is already downloaded, it is not
            downloaded again.
```

File: tests/base/datasets.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    tensor = tensor.clone()
+    mean = torch.as_tensor(mean, dtype=tensor.dtype, device=tensor.device)
+    std = torch.as_tensor(std, dtype=tensor.dtype, device=tensor.device)
+    tensor.sub_(mean).div_(std)
+    return tensor
+
+
+class TestingMNIST(MNIST):
+
+    def __init__(self, root, train=True, normalize=(0.5, 1.0), download=False, num_samples=8000):
```
Comment (Borda) [inline tests/base/datasets.py]: next step would be also select just a subset of digits because we do not have use all 10 in testing...
so with limiting to 5 digits and keeping 2k examples, we may increase back the test's accuracy to 0.5

Comment (awaelchli) [reply tests/base/datasets.py]: good idea. new pr or this one?

File: tests/base/datasets.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    RESOURCES = (
+        "https://pl-public-data.s3.amazonaws.com/MNIST/processed/training.pt",
+        "https://pl-public-data.s3.amazonaws.com/MNIST/processed/test.pt",
+    )
+
+    TRAIN_FILE_NAME = 'training.pt'
+    TEST_FILE_NAME = 'test.pt'
+
+    def __init__(self, root, train=True, normalize=(0.5, 1.0), download=False):
```
Comment (Borda) [inline tests/base/datasets.py]:
```suggestion
    def __init__(self, root: str, train: bool = True, normalize: tuple = (0.5, 1.0), download: bool = False):
```

Comment (awaelchli) [reply tests/base/datasets.py]: should I add types everywhere in MNIST dataset?

File: tests/base/datasets.py (hunk 6 above)
Comment (awaelchli) [reply tests/base/datasets.py]: is accuracy important in tests?

**Comment by awaelchli** [comment]
changelog updated. we also add changes that don't affect users?

File: tests/base/datasets.py (hunk 6 above)
Comment (Borda) [reply tests/base/datasets.py]: yes, acc is used to verify the model is learning something
sure a new PR would be better since it would need to update some more numbers

File: tests/base/datasets.py (hunk 7 above)
Comment (Borda) [reply tests/base/datasets.py]: it would be nice but I was changing this in particular because you had in docs which we are changing

**Comment by Borda** [comment]
> changelog updated. we also log changes that don't affect users?

probably not useful for users but a quite milestone for us so I would mention it there

File: CHANGELOG.md [hunk 8]
```diff
@@ -26,6 +26,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Removed
 
 - Removed duplicated module `pytorch_lightning.utilities.arg_parse` for loading CLI arguments ([#1167](https://github.com/PyTorchLightning/pytorch-lightning/issues/1167))
+- Removed torchvision dependency in tests ([#986](https://github.com/PyTorchLightning/pytorch-lightning/issues/986))
```
Comment (Borda) [inline CHANGELOG.md]:
```suggestion
- Added own MNIST dataset and dropped `torchvision` dependency in tests ([#986](https://github.com/PyTorchLightning/pytorch-lightning/issues/986))
```

File: tests/base/datasets.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        if download:
+            self.download()
+
+        if not self._check_exists():
+            raise RuntimeError('Dataset not found.')
+
+        data_file = self.TRAIN_FILE_NAME if self.train else self.TEST_FILE_NAME
+        self.data, self.targets = torch.load(os.path.join(self.processed_folder, data_file))
+
+    def __getitem__(self, index: int) -> Tuple[Tensor, int]:
```
Comment (Borda) [inline tests/base/datasets.py]: isnt `index`  a keyword or native function, so rather just `idx`?

File: tests/base/datasets.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        return len(self.data)
+
+    @property
+    def processed_folder(self) -> str:
+        return os.path.join(self.root, 'MNIST', 'processed')
+
+    def _check_exists(self) -> bool:
+        train_file = os.path.join(self.processed_folder, self.TRAIN_FILE_NAME)
+        test_file = os.path.join(self.processed_folder, self.TEST_FILE_NAME)
+        return os.path.exists(train_file) and os.path.exists(test_file)
```
Comment (Borda) [inline tests/base/datasets.py]:
```suggestion
        return os.path.isfile(train_file) and os.path.isfile(test_file)
```

File: CHANGELOG.md (hunk 8 above)
Comment (awaelchli) [reply CHANGELOG.md]: should it go to "changed" section?

File: tests/base/datasets.py (hunk 9 above)
Comment (awaelchli) [reply tests/base/datasets.py]: not that I know of. it is left over from torchvision.MNIST when I copied it. I can change it, doesn't hurt.

**Comment by awaelchli** [comment]
The test failed because of this?
```
Expected nothing
Got:
    TRAINS new version available: upgrade to v0.14.1 is recommended!
```
LOL

**Comment by awaelchli** [comment]
@bmartinn is there a way we can suppress such messages during test?

**Comment by Borda** [comment]
> @bmartinn is there a way we can suppress such messages during test?

We know about it... https://github.com/allegroai/trains/issues/119

**Comment by bmartinn** [comment]
Hi @awaelchli , as @Borda mentioned the issue is known, and a fix was already merged into the master branch. 
If CI tests are failing, the easiest  temporarily fix (just for this PR) is to upgrade the *trains* package in the `requirements-extra.txt` to the latest stable release.
There is already a pending [PR](https://github.com/PyTorchLightning/pytorch-lightning/pull/1229) on that as well ...

**Comment by awaelchli** [comment]
@Borda I see KeyBoardInterrupt in CI. Did you have that before?

**Comment by Borda** [comment]
> I see KeyBoardInterrupt in CI. Did you have that before?

not really...

**Comment by Borda** [comment]
@williamFalcon @PyTorchLightning/core-contributors lets get this done...
